crates/kitsune_p2p/kitsune_p2p/src/config.rs

```diff
@@ -0,0 +1,82 @@
+use url2::Url2;
+
+/// Configure the kitsune actor
+#[non_exhaustive]
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+pub struct KitsuneP2pConfig {
+    /// list of sub-transports to be included in this pool
+    pub transport_pool: Vec<TransportConfig>,
+}
+
+impl Default for KitsuneP2pConfig {
+    fn default() -> Self {
+        Self {
+            transport_pool: Vec::new(),
+        }
+    }
+}
+
+/// Configure the network bindings for underlying kitsune transports
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum TransportConfig {
+    /// A transport that uses the QUIC protocol
+    Quic {
+        /// To which network interface / port should we bind?
+        /// Default: "kitsune-quic://0.0.0.0:0".
+        bind_to: Option<Url2>,
+
+        /// If you have port-forwarding set up,
+        /// or wish to apply a vanity domain name,
+        /// you may need to override the local NIC ip.
+        /// Default: None = use NIC ip.
+        override_host: Option<String>,
+
+        /// If you have port-forwarding set up,
+        /// you may need to override the local NIC port.
+        /// Default: None = use NIC port.
+        override_port: Option<u16>,
+    },
+    /// A transport that tls tunnels through a sub-transport (ALPN kitsune-proxy/0)
+    Proxy {
+        /// The 'Proxy' transport is a wrapper around a sub-transport
+        /// We also need to define the sub-transport.
+        sub_transport: Box<TransportConfig>,
+
+        /// Determines whether we wish to:
+        /// - proxy through a remote
+        /// - be a proxy server for others
+        /// - be directly addressable, but not proxy for others
+        proxy_config: ProxyConfig,
+    },
+}
+
+/// Proxy configuration options
+#[derive(Clone, Debug, serde::Serialize, serde::Deserialize, PartialEq)]
+#[serde(tag = "type", rename_all = "snake_case")]
+pub enum ProxyConfig {
+    /// We want to be hosted at a remote proxy location.
+    RemoteProxyClient {
+        /// The remote proxy url to be hosted at
+        proxy_url: Url2,
+    },
+
+    /// We want to be a proxy server for others.
+    /// (We can also deny all proxy requests for something in-between).
+    LocalProxyServer {
+        /// Accept proxy request options
+        /// Default: None = reject all proxy requests
+        proxy_accept_config: Option<ProxyAcceptConfig>,
+    },
+}
+
+/// Whether we are willing to proxy on behalf of others
```

Comment:
Yeah, I have to say I had this question going on too

Reply:
@zippy @Connoropolous -

This is a quirk of simplification. By always wrapping our communications in the proxy protocol, the communication code can be homogeneous. But while individual nodes may want to be directly addressable - they may not want to proxy for other people, hence the ability to reject proxy requests.

If anyone has a better way of saying that in here, feel free to type in a change suggestion.

Reply:
This is just a follow up clarification, but if you wanted to be directly addressable, wouldn't you be able to, and choose to, just TransportConfig::Quic instead of ::Proxy. Or do I have that wrong?

Reply:
This is just a follow up clarification, but if you wanted to be directly addressable, wouldn't you be able to, and choose to, just TransportConfig::Quic instead of ::Proxy. Or do I have that wrong?

Reply:
That's the simplification part - right now it's actually much easier to just make all the network communications homogeneous - i.e. they are all quic wrapped in the proxy protocol. We may want to allow direct quic connections in the future, but it's actually more coding / testing overhead.

Reply:
Would consumers of this code be eligible to use `TransportConfig::Quic` then?

Reply:
Right now nodes must use the exact configuration of Proxy with Quic as the sub transport. The proxy config can be either proxy client through a remote, or local proxy server. If they choose local proxy server, they can choose not to proxy for others.

Reply:
Ok so in other words you set this up as an enum with TransportConfig::Quic on the top level for the sake of a possible future implementation.
From a consumer standpoint, will it error if you currently use it as top-level option, if not maybe it should? or some kind of note in the code or otherwise, saying what you just said "We may want to allow direct quic connections in the future"

Reply:
also a test case just to demonstrate what that behaviour is/would be

Reply:
If it doesn't error, it will create a partition - those that configure quic directly will only be able to make outgoing or receive incoming connections to other nodes that also configure quic directly.

Reply:
Now that I more fully understand the intended use of all this, and it's representation in the config, I'd say this could really lead to some messy and strange and unintended consequences in real world usage, especially as it extrapolates into the future and takes on other potential implementations. I would not make the TransportConfig recursive... I would suggest making a Proxy config variant specifically depend on a separate Config which is a Quic variant, so that you don't have strange things happening. My 'vote' wouldn't count as blocking if I were to request changes, but I figure I should speak openly at least, and let y'all do with it what you will!

Reply:
I don't want to remove functionality just because it's complex.

The goal is to expand connectivity options, like bluetooth, and who knows what else in the future. And some of those options may wish to wrap themselves in proxies, or other hierarchical structures.

This *is* a complex domain which we should totally try to mitigate with clear APIs and good documentation, so I'm all on board with that, but if a particular hApp needs to run in a mode that is different from the norm, I don't feel like we should code it so that is impossible. And Kitsune could be used by other completely non-holochain projects as well.